source/engines/rollEngine.js

```diff
@@ -1,548 +1,771 @@
-			acc.push(cur);
-			return acc;
-		} else { //stop on the first non-roll
-			acc.push('║');
-			return acc;
-		}
-	}, []);
-	rollArray[rollArray.length-1] == '║' ? rollArray.pop() : ()=>{};
-	var textArray = splitInput.reduceRight((acc, cur) => {
-		if (acc[0] == '║' || cur == '') {
-			return acc;
-		} else if (/([0-9]+|[0-9]+d[0-9]+([dk][lh]?[0-9]+)?)([\+\-\*]([0-9]+|[0-9]+d[0-9]+([dk][lh]?[0-9]+)?))*$/.test(cur) || /[0-9]+$/.test(cur) || /\)/.test(cur)) {
-			acc.unshift('║');
-			return acc;
-		} else {
-			acc.unshift(cur);
-			return acc;
-		}
-	}, []);
-	textArray[0] == '║' ? textArray.shift() : ()=>{};
-	return new ResultBundle(rollArray.join(''),textArray.join(' '));
-}
-
-function getRollString(input, frac = false, simple = false) {
-	return parseRoll(input).toString(frac, simple);
-}
-
-module.exports = {parseRoll, getRollString}
+class Result {
+	#result;
+
+	constructor(result) {
+		this.#result = result;
+	}
+
+	getResult() {
+		return this.#result;
+	}
+
+	toString() {
+		return `${this.#result}`;
+	}
+}
+
+class RollResult extends Result {
+	#dieSides;
+
+	constructor(dieSides) {
+		super(Math.floor(Math.random() * dieSides) + 1);
+		this.#dieSides = dieSides;
+	}
+
+	getMaxValue() {
+		return this.#dieSides;
+	}
+
+	toString(frac = false) {
+		return frac ? `${this.getResult()}/${this.getMaxValue()}` : `${this.getResult()}`;
+	}
+}
+
+class StringResult extends Result {
+	#optList;
+
+	constructor(result, optList) {
+		super(result);
+		this.#optList = optList;
+	}
+
+	getMaxValue() {
+		return this.getResult();
+	}
+
+	toString(frac = false) {
+		return frac ? `${this.getResult()} from the list ${this.#optList}` : this.getResult();
+	}
+}
+
+const two_ops = { // Mathematical operations we currently support for rolls that use two arguments
+	ADD: '+',
+	SUBTRACT: '-',
+	MULTIPLY: '*'
+}
+
+const keep_ops = { // Operations for keeping or dropping rolls of a result set
+	KEEP: 'k',
+	KEEP_HIGHEST: 'kh',
+	DROP_LOWEST: 'dl',
+	DROP: 'd',
+	DROP_HIGHEST: 'dh',
+	KEEP_LOWEST: 'kl'
+}
+
+class ResultSet {
+
+	constructor() {}
+
+	static #surroundedByParens(str, symbolPos) {
+		var rightParensUnenclosed = 0,
+			leftParensUnenclosed = 0;
+		var leftSide = str.slice(0,symbolPos),
+			rightSide = str.slice(symbolPos+1,str.length-1);
+		for (var i = leftSide.length - 1; i >= 0 ; i-=1) {
+			if (leftSide[i] == ')' && leftParensUnenclosed <= 0) {
+				leftParensUnenclosed -= 1;
+			} else if (leftSide[i] == '(') {
+				leftParensUnenclosed += 1;
+			}
+		}
+		for (var i = 0; i < rightSide.length; i+=1) {
+			if (rightSide[i] == '(' && rightParensUnenclosed <= 0) {
+				rightParensUnenclosed -= 1;
+			} else if (rightSide[i] == ')') {
+				rightParensUnenclosed += 1;
+			}
+		}
+		return rightParensUnenclosed > 0 && leftParensUnenclosed > 0;
+	}
+
+	static #parseStringList(str) {
+		let strList = [];
+		let newEntry = "";
+		let startQuote = false;
+		let endQuote = false;
+		let spaceCount = 0;
+		for (var i = 0; i < str.length; i++) {

```

Comment:
`startQuote && endQuote` shows up in nearly every case, would probably be convenient to make a helper variable for that.
```suggestion
		for (var i = 0; i < str.length; i++) {
			const bothQuotes = startQuote && endQuote;
```

Reply:
There's maybe an argument that, because we can't have `endQuote == true` unless `startQuote == true`, the structure for this is entirely wrong in the first place. This value is arguably trinary (0, 1, 2), and should be represented as an integer. I might end up doing this for the time being, anyway.